### PR TITLE
feat(parameters): SSMProvider support

### PR DESCRIPTION
**Pull request by dreamorosi**
## Description of your changes

This PR implements the `SSMProvider` class which allows the future Parameters utility to support retrieving parameters from SSM.

The `SSMProvider` supports the following retrieval modes:
- `get` this class method allows customers to retrieve a single parameter by name and allows to specify transform, max age, and decrypt. 
- `getMultiple` this class method allows customers to retrieve multiple parameters by path and allows to specify a shared config object that changes the behavior around transform, max age, and decrypt.
- `getParametersByName` this class method allows customers to retrieve multiple parameters by name. It also allows to specify configs around transform, max age, and decrypt both for every single parameter or using a shared config object. Since the configs can vary, the underlying SDK calls must adapt based on the logic described below.

```sh
                               ┌────────────────────────┐
                           ┌───▶  Decrypt entire batch  │─────┐
                           │   └────────────────────────┘     │     ┌────────────────────┐
                           │                                  ├─────▶ GetParameters API  │
   ┌──────────────────┐    │   ┌────────────────────────┐     │     └────────────────────┘
   │   Split batch    │─── ┼──▶│ No decryption required │─────┘
   └──────────────────┘    │   └────────────────────────┘
                           │                                        ┌────────────────────┐
                           │   ┌────────────────────────┐           │  GetParameter API  │
                           └──▶│Decrypt some but not all│───────────▶────────────────────┤
                               └────────────────────────┘           │ GetParameters API  │
                                                                    └────────────────────┘
```
- `getParameter` this is a standalone function that customers can import & call directly without instantiating the class. Aside from that it's an alias to the class method described above.
- `getParameters` this is a standalone function that customers can import & call directly without instantiating the class. Aside from that it's an alias to the class method described above.
- `getParametersByName` this is a standalone function that customers can import & call directly without instantiating the class. Aside from that it's an alias to the class method described above.

At the time of writing, the implementation in this PR should have feature parity with the implementation [found in Powertools for Python](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/aws_lambda_powertools/utilities/parameters/ssm.py).

### How to verify this change

See the newly added unit test cases & optionally compare also the API surface with the implementation [found in Powertools for Python](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/aws_lambda_powertools/utilities/parameters/ssm.py).

### Related issues, RFCs

**Issue number:** #1176

### PR status

***Is this ready for review?:*** YES  
***Is it a breaking change?:*** NO

## Checklist

- [x] [My changes meet the tenets criteria](https://awslabs.github.io/aws-lambda-powertools-typescript/#tenets)
- [x] I have performed a *self-review* of my own code
- [x] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [ ] I have made corresponding changes to the *documentation*
- [ ] I have made corresponding changes to the *examples*
- [x] My changes generate *no new warnings*
- [x] The *code coverage* hasn't decreased
- [x] I have *added tests* that prove my change is effective and works
- [x] New and existing *unit tests pass* locally and in Github Actions
- [x] Any *dependent changes have been merged and published*
- [x] The PR title follows the [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

- [ ] I have documented the migration process
- [ ] I have added, implemented necessary warnings (if it can live side by side)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/parameters/src/BaseProvider.ts [hunk 1]
```diff
@@ -49,6 +55,11 @@ abstract class BaseProvider implements BaseProviderInterface {
 
     let value;
     try {
+      // Type assertion is needed because the SSM provider has a different signature for the get method
+      if (options && isSSMGetOptionsInterface(options) && options.decrypt) {
+        options.sdkOptions = options.sdkOptions || {};
+        options.sdkOptions.WithDecryption = true;
+      }
```
Comment (buggy) [inline packages/parameters/src/BaseProvider.ts:62]: Would it be better to move this into `_get()` of the `SSMProvider` class so that there's no implementation specific logic in the base class? This comment also applies to `getMultiple()`.

Comment (dreamorosi) [reply packages/parameters/src/BaseProvider.ts:62]: Hi @buggy thanks for the review. 

That's a good point, I'll look into it and see if I can bring that logic into `SSMProvider`.

Comment (dreamorosi) [reply packages/parameters/src/BaseProvider.ts]: I have applied the change you suggested in this commit [e47ddf67](https://github.com/awslabs/aws-lambda-powertools-typescript/pull/1187/commits/e47ddf67f14cde4e41c5bc2dd7a71eedd990d549). 

Thank you!

File: packages/parameters/package.json [hunk 2]
```diff
@@ -51,5 +51,14 @@
     "secrets",
     "serverless",
     "nodejs"
-  ]
+  ],
+  "devDependencies": {
+    "@aws-sdk/client-appconfig": "^3.223.0",
```
[the remark this answers is not in the document]
Comment (dreamorosi) [reply packages/parameters/package.json]: My bad. 

I have removed the non-SSM related dependencies in favor of having them added in their own PRs.

File: packages/parameters/src/SSMProvider.ts [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+    return errors;
+  }
+
+  /**
+   * Split parameters that can be fetched by GetParameters vs GetParameter.
+   */
+  protected static splitBatchAndDecryptParameters(parameters: Record<string, SSMGetParametersByNameOptionsInterface>, configs: SSMGetParametersByNameOptionsInterface): SSMSplitBatchAndDecryptParametersOutputType {
+    const splitParameters: SSMSplitBatchAndDecryptParametersOutputType = {
+      batch: {},
+      decrypt: {},
```
[the remark this answers is not in the document]
Comment (dreamorosi) [reply packages/parameters/src/SSMProvider.ts]: Good callout, I have updated the implementation to use your suggestions.

File: packages/parameters/src/SSMProvider.ts [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+      if (!this.hasKeyExpiredInCache(cacheKey)) {
+        // Since we know the key exists in the cache, we can safely use the non-null assertion operator
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        results.cached[parameterName] = this.store.get(cacheKey)!.value;
+      } else {
+        results.toFetch[parameterName] = parameterOptions;
+      }
+    }
+
+    return results;
```
[the remark this answers is not in the document]
Comment (dreamorosi) [reply packages/parameters/src/SSMProvider.ts]: Good point!

I have changed all the functions that allowed it according to your suggestion and used more explanatory names for the others that didn't.

I have also re-formatted some of the signatures to be on multiple lines since they were very long.

File: packages/parameters/src/types/SSMProvider.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+interface SSMGetParametersByNameOutputInterface {
+  response: Record<string, unknown>
+  errors: string[]
+}
+
+type SSMGetParametersByNameFromCacheOutputType = {
+  cached: Record<string, string | Record<string, unknown>>
+  toFetch: Record<string, SSMGetParametersByNameOptionsInterface>
+} & { [key: string]: SSMGetParametersByNameOptionsInterface };
```
[the remark this answers is not in the document]
Comment (dreamorosi) [reply packages/parameters/src/types/SSMProvider.ts:44]: It was, I have removed them.

File: packages/parameters/src/BaseProvider.ts [hunk 6]
```diff
@@ -35,8 +40,9 @@ abstract class BaseProvider implements BaseProviderInterface {
    * this should be an acceptable tradeoff.
    * 
    * @param {string} name - Parameter name
-   * @param {GetOptionsInterface} options - Options to configure maximum age, trasformation, AWS SDK options, or force fetch
+   * @param {GetOptionsInterface|SSMGetOptionsInterface} options - Options to configure maximum age, trasformation, AWS SDK options, or force fetch
    */
+  public async get(name: string, options?: SSMGetOptionsInterface): Promise<undefined | string | Record<string, unknown>>;
   public async get(name: string, options?: GetOptionsInterface): Promise<undefined | string | Record<string, unknown>> {
```
[the remark this answers is not in the document]
Comment (dreamorosi) [reply packages/parameters/src/BaseProvider.ts]: Yes, it's a function overload which as far as I have been reading, it's purely a TypeScript construct. Since JavaScript proper doesn't really have function overload you are allowed to specify only one implementation (not sure if this answers your question).

I tried union types initially but I ended up with overloading both because of the inputs but also because of the outputs. In this specific case `SSMProvider.get` never returns an `Uint8Array` while other providers do (i.e. `AppConfigProvider`).

If I use the union type the type-checker complains like this:
image

I'm very open to alternatives though. I have to admit that this is very close to the limits of my TS knowledge so if there's a better / more elegant way that allows users to get the right types when using the library I'm open to consider it.

File: packages/parameters/src/SSMProvider.ts [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+      const { response: batchResponse, errors: batchErrors } = await this.getParametersBatchByName(batch, configs.throwOnError, false);
+      
+      response = { ...decryptResponse, ...batchResponse };
+      // Fail-fast disabled, let's aggregate errors under "_errors" key so they can handle gracefully
+      if (!configs.throwOnError) {
+        response[this.errorsKey] = [ ...decryptErrors, ...batchErrors ];
+      }
+    } else {
+      const { response: batchResponse, errors: batchErrors } = await this.getParametersBatchByName(decrypt, configs.throwOnError, true);
+
```
[the remark this answers is not in the document]
Comment (dreamorosi) [reply packages/parameters/src/SSMProvider.ts]: My understanding is that decryption introduces an additional variable in the equation and that this forces us to have to handle parameters that require decryption separately and one-by-one. 

The discussion under the PR that introduced the change in Python ([link](https://github.com/awslabs/aws-lambda-powertools-python/pull/1678)) explains how they settled on this implementation and I think [this message](https://github.com/awslabs/aws-lambda-powertools-python/pull/1678#issuecomment-1305853491), specifically the parts related to KMS, are the reason why we need to treat these ops as separate. 

The TL;DR; is that users might try to retrieve params that have been encrypted with different keys that their function might or might not have rights to use. In order to handle this gracefully we need to isolate these calls.

[the remark this answers is not in the document]
Comment (dreamorosi) [reply packages/parameters/src/SSMProvider.ts]: And to answer your first question about the method being called with a single parameter:
- If the parameter doesn't need encryption, the `if` block will be executed and:
  - `this.getParametersByNameWithDecryptOption` will be called with an empty object, which will cause an immediate return (also empty)
  - `this.getParametersBatchByName` with `false will be called
- If the parameter needs encryption, the `else` block will be executed and `this.getParametersBatchByName` with `true` will be called
